### DISCUSSION TITLE
How to capture an argument to a mock?

### DISCUSSION THREAD

**Discussion by efenderbosch-atg** [spockframework/spock#2050] [Q&A] [answered]
```groovy
        LotFileEntity newFile = null
        1 * lotFileRepo.saveAll({ List<LotFileEntity> l ->
            l[1].guid == existingFile1.guid && l[2].guid == existingFile2.guid && l.size() == 3
        }) >> { List<LotFileEntity> l ->
            List<LotFileEntity> flattened = l.flatten()
            newFile = flattened[0]
            flattened
        }
```

When I run this I get:
```
No such property: newFile for class: org.spockframework.mock.runtime.MockInvocation
groovy.lang.MissingPropertyException: No such property: newFile for class: org.spockframework.mock.runtime.MockInvocation
```

**Comment by Vampire** [comment, marked as answer]
Is that in a `then` block?
If so, then it is moved by the compiler to before the according `when` block.
But the first line of your snippet stays in the `then` block and thus is not defined up where the interaction is placed now.
Wrap both expressions into an `interaction { ... }` block, then they are moved both together and it should work.

**Reply by efenderbosch-atg** [reply]
Thanks. I couldn't quite get it working like that, but declaring `LotFileEntity newFile = null` above the `when` block did work.
